src/ctia/stores/es/query.clj

```diff
@@ -53,3 +55,81 @@
    [{:term {"observable.type" type}}
     {:term {"observable.value" value}}]))
 
+(s/defschema ESQFullTextQuery
+  (st/merge
+   {:query s/Str}
+   (st/optional-keys
+    {:default_operator s/Str
+     :fields [s/Str]})))
+
+(defn- searchable-fields-map-impl
+  [prev m result]
+  (reduce-kv
+   (fn [res key val]
+     (if (= "keyword" (:type val))
+       (let [field-name (some->> val
+                                 :fields
+                                 (filter (fn [[_ v]] (= "text" (:type v))))
+                                 ffirst
+                                 name)
+             path (conj prev (name key))]
+         (if field-name
+           (conj res path field-name)
+           res))
+       (if (map? val)
+         (let [path (if (not= :properties key)
+                      (conj prev (name key))
+                      prev)]
+           (searchable-fields-map-impl path val res))
+         res)))
+   result m))
+
+(s/defn searchable-fields-map :- {s/Str s/Str}
```

Comment:
Do we have a test for that function somewhere? It is a bit hard to review the function without some example of usage 😢

Reply:
In addition, it's sensitive enough to be unit tested and check the edge cases.

Reply:
@DeLaGuardo @ereteog I've added tests: test/ctia/stores/es/query_test.clj

Reply:
It is a bit hard to follow the logic in searchable-fields-map-impl. So far I wrap it in my head as "Walk through data structure until value is a map without `:type "keyword"` collecting on the way all the keys except :properties as a path. Then if the leaf node has at least one field of type text - keep the pair, path and field. Later collection of pair will be turned into hash-map." 
Based on that I wrote alternative implementation:

```clj
(defn- searchable-fields-map-impl
  ([m] (searchable-fields-map-impl [] m))
  ([path m]
   (if (and (map? m) (not= (:type m) "keyword"))
     (let [step-fn (fn [k]
                     (let [path (if (= k :properties) path (conj path k))]
                       (searchable-fields-map-impl path (m k))))]
       (mapcat step-fn (keys m)))
     (let [fields (reduce-kv
                   (fn [acc field {:keys [type]}]
                     (if (= type "text")
                       (conj acc field)
                       acc))
                   []
                   (:fields m))]
       (map #(conj [path] %) fields)))))

(s/defn searchable-fields-map :- {s/Str s/Str}
  "Walks through entity's mapping properties and finds all fields with secondary searchable token.
Returns a map where key is path to a field, and value - path to the nested text token."
  [properties :- (s/maybe {s/Keyword s/Any})]
  (into {}
        (map (fn [[path field]]
               (let [path (str/join "." (map name path))]
                 [path
                  (str/join "." [path (name field)])])))
        (searchable-fields-map-impl properties)))
```

@agzam What do you think?

Reply:
Very nice @DeLaGuardo, your version looks a lot nicer. Thank you for spending time to figure it out. Switched to your implementation.